### DISCUSSION TITLE
Does not work with (treesit-defun-at-point)

### DISCUSSION THREAD

**Discussion by drshapeless** [minad/tempel#118] [Q&A] [answered]
I was adding a template to insert the current function name in my snippet, but it doesn't work.

```
go-ts-mode

(ef
 > "if err != nil {" n
 > "app.error(\"" (treesit-defun-name (treesit-defun-at-point)) "\")" n
 > "}")

(fn (treesit-defun-name (treesit-defun-at-point)))
```
Let say I have a function named bar in my Go code.

Expanding fn will give me the funtion name, bar.

Expanding ef will weirdly give me an empty string in the same function.

I tried to recreate my intention using simple elisp function.
```
(defun drsl/go-debug ()
  (interactive)
  (treesit-indent)
  (insert "if err != nil {")
  (newline)
  (treesit-indent)
  (insert "app.error(\"" (treesit-defun-name (treesit-defun-at-point)) "\")")
  (newline)
  (treesit-indent)
  (insert "}")
  )
```

It also doesn't work, because the function `(treesit-defun-name (treesit-defun-at-point))` return nil after any insertion. 

My workaround is using a let statement to enclose the whole thing and get the name first.
```
(defun drsl/go-debug ()
  (interactive)
  (let ((NAME (treesit-defun-name (treesit-defun-at-point))))
    (treesit-indent)
    (insert "if err != nil {")
    (newline)
    (treesit-indent)
    (insert "app.error(\"" NAME "\")")
    (newline)
    (treesit-indent)
    (insert "}")))
```

Now the thing works. If I want it to work with tempel, now I have to change my ef snippet into this.
```
(ef (drsl/go-debug))
```

How could I do it with tempel without writing custom function? I suspect tempel suffers from the same bug of my first iteration of the function.

**Comment by minad** [comment, marked as answer]
In order for this to work `(treesit-defun-name (treesit-defun-at-point))` should be evaluated first and bound to a variable `name`. You can do this using `(p FORM <NAME> <NOINSERT>)`:

`(p (treesit-defun-name (treesit-defun-at-point)) name t)`

The variable can then be used later. See also https://github.com/minad/tempel#template-syntax for the supported syntax. In case you want to reuse a pattern often, you can define custom elements: https://github.com/minad/tempel#defining-custom-elements.
